pkg/controller/controller_flow_instance_test.go

```diff
@@ -162,3 +164,755 @@ func TestServiceInstanceDeleteWithAsyncUpdateInProgress(t *testing.T) {
 		})
 	}
 }
+
+// TestCreateServiceInstanceFailsWithNonexistentPlan tests creating a ServiceInstance whose ServicePlan
+// does not exist
+func TestCreateServiceInstanceFailsWithNonexistentPlan(t *testing.T) {
+	// GIVEN
+	ct := newControllerTest(t)
+	defer ct.TearDown()
+
+	ct.SetupEmptyPlanListForOSBClient()
+	require.NoError(t, ct.CreateSimpleClusterServiceBroker())
+	require.NoError(t, ct.WaitForReadyBroker())
+
+	// WHEN
+	require.NoError(t, ct.CreateServiceInstance())
+
+	// THEN
+	condition := v1beta1.ServiceInstanceCondition{
+		Type:   v1beta1.ServiceInstanceConditionReady,
+		Status: v1beta1.ConditionFalse,
+		Reason: "ReferencesNonexistentServicePlan",
+	}
+	require.NoError(t, ct.WaitForInstanceCondition(condition))
+}
+
+// TestCreateServiceInstanceNonExistentClusterServiceBroker tests creating a
+// ServiceInstance whose broker does not exist.
+func TestCreateServiceInstanceNonExistentClusterServiceBroker(t *testing.T) {
+	// GIVEN
+	ct := newControllerTest(t)
+	defer ct.TearDown()
+
+	require.NoError(t, ct.CreateClusterServiceClass())
+	require.NoError(t, ct.WaitForClusterServiceClass())
+
+	require.NoError(t, ct.CreateClusterServicePlan())
+	require.NoError(t, ct.WaitForClusterServicePlan())
+
+	// WHEN
+	require.NoError(t, ct.CreateServiceInstance())
+
+	// THEN
+	condition := v1beta1.ServiceInstanceCondition{
+		Type:   v1beta1.ServiceInstanceConditionReady,
+		Status: v1beta1.ConditionFalse,
+		Reason: "ReferencesNonexistentBroker",
+	}
+	require.NoError(t, ct.WaitForInstanceCondition(condition))
+}
+
+// TestCreateServiceInstanceNonExistentClusterServiceClassOrPlan tests that a ServiceInstance gets
+// a Failed condition when the service class or service plan it references does not exist.
+func TestCreateServiceInstanceNonExistentClusterServiceClassOrPlan(t *testing.T) {
+	// TODO: cannot rewrite tests because fakeClient does not support `FieldSelector` during filtering ServiceInstance list
```

Comment:
add info that it should be added after merging the CRDs

Reply:
Done.